### ISSUE TITLE
Add Recurrent Layers

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/axon#24]
Require solution for managing layers that maintain state / return multiple outputs

- [ ] gru
- [ ] lstm
- [ ] conv_lstm

**Comment by nyo16** [comment]
if i have some time this weekend i want to tackle RNN | LSTM and write an example for simple sequence prediction. 
Managing the state will be an interesting problem :)

**Comment by batate** [comment]
That sounds great to me. I will be watching this one.

**Comment by seanmor5** [comment]
Hi @nyo16 sounds good. I haven't thought much about how I want this API to look but I would love to see what you come up with. Note I've also excluded RNN as you could call any of GRU, LSTM, etc. an RNN. It may be best instead to introduce this as SimpleRNN like Keras. I will think on this issue as well this weekend and dig out a few papers / other frameworks to see what they are doing so I can provide better feedback and direction.

**Comment by seanmor5** [comment]
I've done a survey of recurrent layers in practice and in theory. For reference, I'm following along mostly with discussions in the following books:

- [Supervised Sequence Labeling](https://www.cs.toronto.edu/~graves/preprint.pdf)
- [Deep Learning Book Chapter 10: Sequence Modeling](https://www.deeplearningbook.org/contents/rnn.html)

As for practical implementations, I've gotten a pretty decent understanding of [Haiku](https://github.com/deepmind/dm-haiku/blob/master/haiku/_src/recurrent.py) and [Flax](https://github.com/google/flax/blob/master/flax/linen/recurrent.py) implementations of RNNs.

A general RNN essentially takes a hidden state and an input and transforms them across timesteps starting from an initial state. In most implementations, the atomic RNN operation is referred to as a cell (e.g. GRUCell, LSTMCell, etc.). We can implement these today without any issues using `defn`. However, cells only apply the operation across a single timestep. What we really want is to apply the cells across entire sequences by "unrolling" the input sequence.

In order to "unroll" the RNNs, we have two options (I'm using the terminology put forth in Haiku here):

1) Static Unroll - A static unroll applies a cell to each timestep of an input sequence before jitting the expression. Basically, we do:

```elixir
defn static_unroll(sequence, state, cell_fn) do
  transform({sequence, cell_fn}, fn {sequence, cell_fn} ->
    {timesteps} = Nx.shape(sequence)
    for i <- 0..timesteps - 1, reduce: state do
      cell_fn.(sequence[i], state)
    end
  end)
end
```

So we yield a very large expression by slicing and applying the cell at each timestep. This has the advantage in that we can do it right now without any limitations in Nx. The major disadvantage is that this yields REALLY large expressions and will ultimately lead to REALLY long compilation and runtime. It's not efficient at all for large sequences.

2) Dynamic Unroll - A dynamic unroll applies a cell to each timestep of an input sequence as a part of the expression using a reduction. In Haiku, this is implemented using [jax.lax.scan](https://github.com/google/jax/blob/master/jax/_src/lax/control_flow.py#L1141). In this case, the unrolling is built into the jitted expression so it's much more efficient. Unfortunately, that means we're blocked on: https://github.com/elixir-nx/nx/issues/122

Moving forward to progress on this issue, I'm proposing the following roadmap:

1) Implement `gru_cell`, `lstm_cell`, `conv_lstm_cell` in `Axon.Layers`. These are relatively straightforward implementations. They should all return a transformed output and a hidden state as a tuple.

2) Implement `static_unroll`. As for where `static_unroll` belongs, it may make sense to put it into it's own `Axon.Recurrent` namespace and move each of the cell implementations there, but perhaps `static_unroll` can be implemented generally enough to be useful elsewhere.

3) Implement `gru_cell`, `lstm_cell`, `conv_lstm_cell` in `Axon` as a part of the high-level API. This will require some extra work considering each layer returns a tuple and it might be desirable to do something like:

```elixir
{x, state} = Axon.lstm_cell(inp)
{Axon.dense(x), Axon.lstm_cell(x, state: state)}
```

Or at least to work with the state and output separately.

4) Implement `gru`, `lstm`, `conv_lstm` in `Axon` as a part of the high-level API with the option to choose an unrolling strategy. Obviously only static unrolling will be supported to start, but once we support while loops in Nx, we will make that the default and potentially phase out the option to statically unroll the RNN.
